src/librawspeed/decompressors/LJpegDecompressor.cpp

```diff
@@ -217,8 +249,37 @@ template <int N_COMP, bool WeirdWidth> void LJpegDecompressor::decodeN() {
       assert(trailingPixels < N_COMP);
       unsigned c = 0;
       for (; c < trailingPixels; ++c) {
-        pred[c] = uint16_t(pred[c] + ht[c]->decodeDifference(bitStream));
-        img(row, col + c) = pred[c];
+        // Continue predictor update skipped at last full block
+        int predA = img(row, col - N_COMP + c);
+        int predB = predMode > 1 ? img(row - 1, col + c) : 0;
+        int predC = predMode > 1 ? img(row - 1, col - N_COMP + c) : 0;
+        switch (predMode) {
+        case 1:
+          pred[c] = uint16_t(predA);
+          break;
+        case 2:
+          pred[c] = uint16_t(predB);
+          break;
+        case 3:
+          pred[c] = uint16_t(predC);
+          break;
+        case 4:
+          pred[c] = uint16_t(predA + predB - predC);
+          break;
+        case 5:
+          pred[c] = uint16_t(predA + ((predB - predC) >> 1));
+          break;
+        case 6:
+          pred[c] = uint16_t(predB + ((predA - predC) >> 1));
+          break;
+        case 7:
+          pred[c] = uint16_t((predA + predB) >> 1);
```

Comment:
When swiching to int32_t, predictor 7 is not calculated by arithmetic right shift, but with division by 2 (see ITU 81, table H.1).

Reply:
I was split about this, as indeed the spec does not have same footnote for pred 7 as it does for 5 and 6... What's the difference, it is integer division so there is no rounding, no?

Reply:
For unsigned int, yes. But for signed int, division and shift is not the same. And the prediction formulas are written for signed integers.

Reply:
For predictor 7, you can't have a negative sum, both quantities (previous samples) are positive.

Reply:
I really wished they were clearer on this in the spec...

Reply:
You are right, A and B could not be negative, my fault. But I remembered that part of the spec and the *wtf* moment while reading it.